src/box/xlog.c

```diff
@@ -1158,43 +1158,73 @@ xlog_tx_write_zstd(struct xlog *log)
 	}
 	uint32_t crc32c = 0;
 	struct iovec *iov;
-	/* 3 is compression level. */
-	ZSTD_compressBegin(log->zctx, 3);
+	size_t rc =
+		ZSTD_CCtx_setParameter(log->zctx, ZSTD_c_compressionLevel, 3);
+	if (ZSTD_isError(rc)) {
+		diag_set(ClientError, ER_COMPRESSION, ZSTD_getErrorName(rc));
+		goto error;
+	}
 	size_t offset = XLOG_FIXHEADER_SIZE;
 	for (iov = log->obuf.iov; iov->iov_len; ++iov) {
-		/* Estimate max output buffer size. */
-		size_t zmax_size = ZSTD_compressBound(iov->iov_len - offset);
-		/* Allocate a destination buffer. */
-		void *zdst = obuf_reserve(&log->zbuf, zmax_size);
-		if (!zdst) {
-			diag_set(OutOfMemory, zmax_size, "runtime arena",
-				  "compression buffer");
-			goto error;
-		}
-		size_t (*fcompress)(ZSTD_CCtx *, void *, size_t,
-				    const void *, size_t);
+		ZSTD_outBuffer zobuf = {
+			.dst = NULL,
+			.size = 0,
+			.pos = 0,
+		};
+		ZSTD_inBuffer zibuf = {
+			.src = (char *)iov->iov_base + offset,
+			.size = iov->iov_len - offset,
+			.pos = 0,
+		};
 		/*
 		 * If it's the last iov or the last
 		 * log has 0 bytes, end the stream.
 		 */
-		if (iov == log->obuf.iov + log->obuf.pos ||
-		    !(iov + 1)->iov_len) {
-			fcompress = ZSTD_compressEnd;
-		} else {
-			fcompress = ZSTD_compressContinue;
-		}
-		size_t zsize = fcompress(log->zctx, zdst, zmax_size,
-					 (char *)iov->iov_base + offset,
-					 iov->iov_len - offset);
-		if (ZSTD_isError(zsize)) {
-			diag_set(ClientError, ER_COMPRESSION,
-				 ZSTD_getErrorName(zsize));
-			goto error;
+		ZSTD_EndDirective mode =
+			iov == log->obuf.iov + log->obuf.pos ||
+			(iov + 1)->iov_len == 0 ? ZSTD_e_end : ZSTD_e_continue;
+		while (true) {
+			if (zobuf.pos == zobuf.size) {
```

Comment:
> output->pos and input->pos will be updated. They are guaranteed to remain below their respective limit.

After reading docs, I'm not sure that the output buffer is guaranteed to reach its limit. Also, [here](https://github.com/facebook/zstd/blob/dev/examples/streaming_compression.c) authors rely on return value and position of input buffer, not position of output buffer. I think we should rewrite this logic.

Reply:
> ZSTD_e_continue is guaranteed to make some forward progress when called, but doesn't guarantee maximal forward progress. [^1]

> The function nonetheless guarantees forward progress : it will return only after it reads or write at least 1+ byte. [^2]

[^1]: https://github.com/facebook/zstd/blob/118200f7b95deaf38b3368cb445a564f187da1a2/lib/zstd.h#L706-L707

[^2]: https://github.com/facebook/zstd/blob/118200f7b95deaf38b3368cb445a564f187da1a2/lib/zstd.h#L770

Reply:
Removed this `if`.